### ISSUE TITLE
Support for upsert?

### ISSUE THREAD

**Issue by GabeKuslansky** [IRCraziestTaxi/typeorm-linq-repository#69]
Is it possible to run an upsert query with this repository? I'd like to create an entity or update it if it doesn't exist without having to run two separate queries (ex: select, then update or insert).

**Comment by IRCraziestTaxi** [comment]
Hi @GabeKuslansky! Sorry for the delayed reply.

I looked into it a little bit and it should be fairly easy to implement. As soon as I get some time (should be able to work on it within the next week or so), I will implement it and get back to you.

Thanks for using typeorm-linq-repository! I always appreciate folks using it and giving feedback to make it better. :)

**Comment by GabeKuslansky** [comment]
@IRCraziestTaxi no worries. I have a short background in C# and miss its luxuries so this repository has been a jackpot of a find for me. Thanks for looking into it.

This might have to be a separate issue/feature request, but if you're planning on implementing upsert, there is another type of functionality that I think would be useful. Being able to update an entity by one or more properties instead of needing to pass the entity object 
```
repo.update(Entity)
```
vs.
```
 UserRepository
     .update(e => e.val)
     .toValue(val);
```
 Not sure how you'd want to go about chaining update props.. perhaps something along the lines of a subsequent .update() or .thenUpdate()
```
UserRepository
    .update(e => e.prop1)
    .toValue(val1)
    .thenUpdate(e => e.prop2)
    .toValue(val2);
```
It's currently a bit awkward to construct an entire entity with all of its properties (either grabbing it from the db or manually mapping) just to change a value for one column. Thank you again for maintaining this project! Typesafe and intellisense producing queries keep me sane.

**Comment by GabeKuslansky** [comment]
There's a situation for me where calling .update() actually generates an "INSERT INTO" query and causes unique constraint errors from the db. Any updates on upsert support?

**Comment by IRCraziestTaxi** [comment]
Hey @GabeKuslansky, unfortunately I have been slammed with work for a while (only recently made an update to this lib that I needed for work) and haven't looked too much into this.

I did just look into the source and see that this lib's `update` method calls TypeORM's `save` method, which it probably should not do since it is basically an upsert.

I'll look into separating those methods as soon as I can find the time.

**Comment by IRCraziestTaxi** [comment]
@GabeKuslansky Would you mind sharing a repro or at least a few code snippets to demonstrate how you got the unique constraint errors? I tested out upserting single entities as well as collections with the `update` method and got expected upsert behavior with no errors, but my sample code was very basic.

**Comment by GabeKuslansky** [comment]
@IRCraziestTaxi 
Assume we have an entity that has a One-to-One relationship with a value object (Say, player and their settings).

If we try to cascade update the settings by calling .update(), the repository will generate an insert into query. This is a problem because my Settings entity already exists in the database with default values, so the unique constraint is enforced...
```sql
query failed: INSERT INTO "player_settings"("id", "isNotifiableByMention", "volume", "favoriteClasses", "addToFavoritesAfterMatch", "addToFavoritesOnLogin", "audioCuesEnabled", "voicepack", "colorOfNameInChat", "playerSteamid") VALUES (NULL, ?, ?, ?, ?, ?, ?, ?, ?, ?) -- PARAMETERS: [0,501,"Medic",0,0,0,"default","default","76561198119135809"]
Error: SQLITE_CONSTRAINT: UNIQUE constraint failed: player_settings.playerSteamid
```

```ts
async updateSettings(steamid: SteamID, settings: PlayerSettings) {
const playerRepository = new LinqRepository(Player);
const player = await this.getPlayer(steamid);
player.settings = settings;
await playerRepository.update(player);
}
```

Here's the Player's relationship
```ts
@OneToOne(type => PlayerSettings, settings => settings.player, {
		cascade: true
	})
```

Here's the PlayerSettings relationship
```ts
@OneToOne(type => Player, player => player.settings)
```

**Comment by GabeKuslansky** [comment]
Update.. was able to fix the issue this way
```ts
const player = await playerRepository.getOne().where(p => p.steamid).equal(steamid).include(p => p.settings);
		const settingsId = player.settings.id;
		player.settings = settings;
		player.settings.id = settingsId;
		await playerRepository.update(player);
```

My getPlayer method doesn't run a .include(player => player.settings) method (which would include the primary key id of settings table), so I manually tied together the relationships.

**Comment by IRCraziestTaxi** [comment]
Sorry I have been MIA for a few days. I've been super busy with work and other stuff.

My first question would be: does the incoming `settings: PlayerSettings` in your `updateSettings` method have the ID of the existing settings?

If not, then there lies the problem because I'm _fairly sure_ TypeORM will perform an update rather than an insert if the property on the object to update that corresponds to the PK matches the existing object. Without that ID, it has no way of knowing that it is an existing object and will try to perform an insert (which is expected behavior IMO).

Based on the fact that your change to set the ID on the incoming `settings` does cause a correct update rather than an insert, I am assuming that is the case - the ID is not on the incoming on which to base the update.

If, however, the incoming object does have its ID, let me know and I will investigate further.

Also, this could give you a much cleaner solution:

I always make a point to explicitly specify the FK property of a relationship because it makes `join`s, `where`s, etc. SO much easier, and would very easily clear up this problem.

For instance:

```ts
// Player
@OneToOne(type => PlayerSettings, settings => settings.player, {
    cascade: true
})
// This ties the FK property below to this relationship
@JoinColumn({ name: "settingsId" })
public settings: PlayerSettings;

@Column()
public settingsId: ...; // string or number, which ever you use

// PlayerSettings
@OneToOne(type => Player, player => player.settings)
public player: Player;

// Do not set an explicit FK in this entity
// because I know TypeORM doesn't handle OneToOne relationships
// with a JoinColumn on both sides of the relationship very well.
```

This way, your update looks like:

```ts
const playerRepository = new LinqRepository(Player);
const player = await playerRepository
    .getOne()
    .where(p => p.steamid)
    .equal(steamid);
player.settings = settings;
player.settings.id = player.settingsId;
await playerRepository.update(player);
```

If that appeals to you, try it out and let me know if it works. It works fantastically with ManyToOne relationships, but I think TypeORM has kinda squirrelly behavior with OneToOne relationships with a JoinColumn decorator. I could be thinking of something else, though.

**Comment by IRCraziestTaxi** [comment]
Hey @GabeKuslansky, were you able to ponder my previous comment? I'd like to close this issue if possible.

Also, I feel like this was previously discussed, although I cannot find it: I looked into it and under the hood, this lib was always calling TypeORM's `save` method for both `create` and `update`; the only difference is that `create` sets the ID on the object to `undefined` (unless prevented via a repository option) to ensure a new entity is created.

Therefore, I created an `upsert` method since that is really what `update` does by calling TypeORM's `save` and simply call that method from `update`; therefore, `update` is now essentially an alias for `upsert`. I left the `update` method in rather than removing it in order to not introduce tons of breaking changes.

Let me know if you have any other concerns; otherwise, I'll close this issue in a few days or so.
